terraform/groups/provisioning-role/profiles/development-eu-west-2/vars

```diff
@@ -0,0 +1,16 @@
+# IAM roles
+iam_roles = {
+  payment-reconciler-provisioning-role = "payment-reconciler-provisioning-role"
+}
+
+# IAM role policies
+iam_role_policies = {
+  payment-reconciler-terraform-plan = "payment-reconciler-provisioning-role"
+  payment-reconciler-terraform-apply = "payment-reconciler-provisioning-role"
+  payment-reconciler-terraform-destroy = "payment-reconciler-provisioning-role"
+}
```

Comment:
Is this the best way to do this? Seems like a lot of duplication

Reply:
The reason I've done it this way (comes from the original spike and I've retained it) is that you can provide a list of roles, and specify which policies attach to which roles. In the example here there's just one role (and all policies attach to it) - but this repo will be used as an example, so kept this in. I think it just looks like duplication because it's three policies (the keys) and they all attach to the same role (the values) but in other cases there could be more than one role. So in that sense, I'm not sure it as duplicated as it first seems.
It's used here:

> resource "aws_iam_role_policy_attachment" "iam_role_policies" {
  for_each   = var.iam_role_policies
  role       = each.value
  policy_arn = lookup(var.iam_policies_arns,each.key)
}

Reply:
The choice of data structures here is forcing the duplication that @clementss pointed out. It should be possible to express the many-to-one relationship between policies and roles within the data structure itself. Any data structure can be transformed later to suit iteration of resources, and generally we should be aiming for the simplest abstraction at this point to simplify the definition and management of environment-specific configuration.

Rather than use distinct data structures, consider using a single map:

```
variable "role_policy_mappings" {
  type = map(object({
    aws_managed_policies = list(string)
    customer_managed_policies = list(string)
  }))
  description = "A map of objects comprising role policy mappings for customer and AWS managed policies"
}
```

Here's an example of a populated data structure:

```
role_policy_mappings = {
  "payment-reconciler-provisioning-role" = {
    "aws_managed_policies": [
      "AWSLambdaReadOnlyAccess"
    ],
    "customer_managed_policies": [
      "payment-reconciler-terraform-plan"
      "payment-reconciler-terraform-apply"
      "payment-reconciler-terraform-destroy"
    ]
  }
}
```

The data structure used to define the mappings can then be transformed into a structure suitable for iteration of resources:

```
locals {
  role_policy_mappings = flatten([
    for role, policy_mappings in var.role_policy_mappings : [
      for policy in policy_mappings.customer_managed_policies : {
        role_name = role
        policy_arn = lookup(var.iam_policies_arns, policy)
      }
    ]
  ])
}
```

The transformed data structure can be used when creating resources:

```
resource "aws_iam_role" "iam_role" {
  for_each           = local.role_policy_mappings
  name               = each.value.role_name
  assume_role_policy = data.aws_iam_policy_document.trust_relationship.json
}
```

```
resource "aws_iam_role_policy_attachment" "iam_role_policies" {
  for_each   = local.role_policy_mappings
  role       = each.value.role_name
  policy_arn = each.value.policy_arn
}
```

Reply:
The choice of data structures here is forcing the duplication that @clementss pointed out. It should be possible to express the many-to-one relationship between policies and roles within the data structure itself. Any data structure can be transformed later to suit iteration of resources, and generally we should be aiming for the simplest abstraction at this point to simplify the definition and management of environment-specific configuration.

Rather than use distinct data structures, consider using a single map:

```
variable "role_policy_mappings" {
  type = map(object({
    aws_managed_policies = list(string)
    customer_managed_policies = list(string)
  }))
  description = "A map of objects comprising role policy mappings for customer and AWS managed policies"
}
```

Here's an example of a populated data structure:

```
role_policy_mappings = {
  "payment-reconciler-provisioning-role" = {
    "aws_managed_policies": [
      "AWSLambdaReadOnlyAccess"
    ],
    "customer_managed_policies": [
      "payment-reconciler-terraform-plan"
      "payment-reconciler-terraform-apply"
      "payment-reconciler-terraform-destroy"
    ]
  }
}
```

The data structure used to define the mappings can then be transformed into a structure suitable for iteration of resources:

```
locals {
  role_policy_mappings = flatten([
    for role, policy_mappings in var.role_policy_mappings : [
      for policy in policy_mappings.customer_managed_policies : {
        role_name = role
        policy_arn = lookup(var.iam_policies_arns, policy)
      }
    ]
  ])
}
```

The transformed data structure can be used when creating resources:

```
resource "aws_iam_role" "iam_role" {
  for_each           = local.role_policy_mappings
  name               = each.value.role_name
  assume_role_policy = data.aws_iam_policy_document.trust_relationship.json
}
```

```
resource "aws_iam_role_policy_attachment" "iam_role_policies" {
  for_each   = local.role_policy_mappings
  role       = each.value.role_name
  policy_arn = each.value.policy_arn
}
```

Reply:
Yep, I see what you mean @marcransome and @clementss 
I did wonder about creating an object for this earlier in its development (I'd been looking at the example used for parameters) but wasn't convinced the extra code complexity was worth it.
I can see the advantage of the object, but I guess the extra nested loops just restructure it into a many-to-many list structure (with duplication) anyway. So perhaps maybe clearer and more readable to just create it in that format in the first place?
What do you think?